Feature/sign in

SangJLee1103 opened the pull request:
## 작업 내용

- [ ] 구글 로그인 구현
- [ ] 로그인 에러 처리 구현
- [ ] 로그인 UI 리팩토링


## 화면 (Optional)

<img width="300" height="600" alt="KakaoTalk_Photo_2025-08-19-22-02-32" src="https://github.com/user-attachments/assets/959a5033-3106-4670-97ed-bbaa857b903c" />




<br/><br/><br/>
## 논의 해봤으면 좋으점 (Optional)
- SNS 로그인 시에 SNS 토큰을 못받아 온다거나 도중에 오류가 나는 경우에 대한 처리에 대해 논의 해야 합니다!

SangJLee1103 committed:
[ADD] pod file 추가

SangJLee1103 committed:
[REFACTOR] SignIn UI 리팩토링

SangJLee1103 committed:
[FEAT] 구글 로그인 구현 및 바인딩

dlrjswns commented on GraceLog/Source/Presentation/SignInScene/ViewController/SignInViewController.swift:
```diff
@@ -46,12 +46,10 @@ final class SignInViewController: UIViewController {
     
     private let leftLine = UIView().then {
         $0.backgroundColor = .themeColor
-        $0.setDimensions(width: 78, height: 1)
     }
     
     private let rightLine = UIView().then {
         $0.backgroundColor = .themeColor
-        $0.setDimensions(width: 78, height: 1)
     }
```
이 부분을 2개 lineView로 두는것보다 1개 lineView에 중앙에 라벨 배치시켜서 원하는만큼 띄워주는게 더 좋을꺼같아요 !!

dlrjswns commented on GraceLog/Source/Presentation/SignInScene/ViewController/SignInViewController.swift:
```diff
@@ -106,25 +114,27 @@ final class SignInViewController: UIViewController {
-        loginStack.distribution = .fillEqually
-        loginStack.spacing = 27
-        
+        [startLabel, leftLine, rightLine, appleLoginButton, googleLoginButton, kakaoLoginButton].forEach {
+            $0.alpha = 0
+        }
+    }
```
공간여백을 없애주세요 !! UI별로 일일이 alpha값을 0으로 바꾸는 방법말고 슈퍼뷰자체를 이용하여 간단하게 할 수는 없을까요 ??

dlrjswns commented on GraceLog/Source/Domain/Usecase/SignIn/Protocol/SignInUseCase.swift:
```diff
@@ -11,7 +11,8 @@ import RxRelay
 protocol SignInUseCase {
     var isSuccessSignIn: BehaviorRelay<Bool> { get }
-    var user: BehaviorRelay<GraceLogUser?> { get } 
+    var user: BehaviorRelay<GraceLogUser?> { get }
```
로그인 성공할때에 유저정보는 UserManager에 적용될꺼라서 굳이 유스케이스에서 방출할 프로퍼티는 필요없을꺼 같아요 !!

SangJLee1103 replied:
음 이걸 둔 이유는 

```
func transform(mutation: Observable<Mutation>) -> Observable<Mutation> {
        let userMutation = usecase.user
            .compactMap { $0 }
            .do(onNext: { [weak self] _ in
                self?.coordinator?.didSignIn()
            })
            .flatMap { _ in Observable<Mutation>.empty() }
        
        let errorMutation = usecase.error
            .map { error in Mutation.setError(error) }
        
        return .merge(mutation, userMutation, errorMutation)
    }
```

여기서 유즈케이스에서 방출한 프로퍼티를 통해 화면전환을 하려고 한건데 혹시 더 좋은 방법이 있을까요??

dlrjswns replied:
user에 대한 정보가 메인탭화면으로 이동할대에 필요가 없기때문에 지워서 의도에 맞는 Observable을 방출하는게 맞다 생각해요

그리고 지금 올려주신 코드 보니까 느낀건데 coordinator는 화면전환역할맡고있는데 didSignIn이라는 네이밍이 아니라 어떤 화면으로 이동하는 함수인지를 판별할 수 있는 네이밍이 좋을꺼같아요

dlrjswns commented on GraceLog/Source/Presentation/SignInScene/ViewController/SignInViewController.swift:
```diff
@@ -183,55 +183,40 @@ extension SignInViewController: View {
-            .disposed(by: disposeBag)
-        
         reactor.state
             .map { $0.isLoading }
             .bind(onNext: { [weak self] isLoading in
                 if isLoading {
-                    self?.activityIndicator.isHidden = false
                     self?.activityIndicator.startAnimating()
                 } else {
                     self?.activityIndicator.stopAnimating()
```
아마 인디케이터도 바로 Bool값을 바인딩할 수 있는 Observable이 있을텐데 이를 이용해주세요

dlrjswns replied:
또한 UI에 바인딩될 수 있는 값은 명시적으로 메인스레드에서 동작할 수 있는 환경으로 만들어주세요

SangJLee1103 replied:
네 이 부분은 UIActivityIndicatorView를 사용하면 Bool값을 바인딩 할 수 있는 isAnimating값이 있는데 현재는 NVActivityIndicatorView라는 라이브러리를 쓰고 있어서 이건 Custom으로 제가 Extension으로 구현해서 적용할게요. 

```
    /// Current status of animation, read-only.
    @available(*, deprecated)
    public var animating: Bool { return isAnimating }

    /// Current status of animation, read-only.
    private(set) public var isAnimating: Bool = false
```

dlrjswns commented on GraceLog/Source/Presentation/SignInScene/ViewController/SignInViewController.swift:
```diff
@@ -262,16 +247,34 @@ extension SignInViewController {
+                    return
+                }
+                
+                self?.reactor?.action.onNext(.googleLogin(token: token))
```
여기서 보내는 토큰값이 저희 서버에서 식별할 수 있는 토큰값이 맞는지 공식링크를 첨부해주세요

SangJLee1103 replied:
공식 링크라는게 뭘까요??.. kakao, google Dev에 있는 개발 관련 페이지 링크인가요?

dlrjswns replied:
@SangJLee1103 아마 저 토큰값이 정확한 사용자의 식별자로써 사용되야할텐데 개발문서에 저 값을 이용하는게 옳은 방법인지를 판별하는 페이지가 있을꺼잖아요 ? 그거 코드상으로 참조링크주석으로 올려주시면 됩니다.

SangJLee1103 replied:
[https://developers.google.com/identity/sign-in/ios/backend-auth?hl=ko](url) 네 여기 있습니다~!

dlrjswns replied:
@SangJLee1103 아뇨아뇨 저희 로그인로직상에서 어떤 공식문서를 사용했는지 적어주는게 좋을꺼같아요 !!

dlrjswns reviewed (changes requested):
@SangJLee1103 수고많으셨습니다, 개인적으로 SignInViewController에서 사용하는 copyrightLabel만 따로 레이아웃을 잡기보다 공간여백을 차지할 수 있는 UIView를 넣고 커지게끔 구현하면 훨씬 깔끔해질꺼같아요

SangJLee1103 committed:
[REFACTOR] 로그인 UI 애니메이션에 대한 리팩토링

SangJLee1103 committed:
[REFACTOR] NVActivityIndicatorView 커스텀 바인딩 구현 및 코드 리팩토링

SangJLee1103 committed:
[FIX] 로그인 성공시 유저 저장에 대한 프로퍼티, 화면전환 함수 네이밍 수정

dlrjswns commented on GraceLog/Source/Application/GraceLogAppCoordinator.swift:
```diff
@@ -45,7 +45,7 @@ final class GraceLogAppCoordinator: NavigationCoordinator {
 
 extension GraceLogAppCoordinator: SignInCoordinatorDelegate {
-    func didSignIn(_ coordinator: SignInCoordinator) {
+    func showHomeFlow(_ coordinator: SignInCoordinator) {
```
홈화면으로 이동하는게 아니라 메인탭화면으로 이동하는 함수라서 네이밍 수정이 필요해보여요 !

dlrjswns commented on GraceLog/Source/Domain/Usecase/SignIn/DefaultSignInUseCase.swift:
```diff
@@ -11,7 +11,7 @@ import RxRelay
 final class DefaultSignInUseCase: SignInUseCase {
     var isSuccessSignIn = BehaviorRelay<Bool>(value: false)
-    var user = BehaviorRelay<GraceLogUser?>(value: nil)
+    var isSuccessFetchUser = BehaviorRelay<Bool>(value: false)
```
BehaviorRelay를 쓰면 아직 사용자에 대한 정보를 올바르게 받아왔는지 미정임에도 false가 떨어져요, 초기값없이 변경해주세요

SangJLee1103 committed:
[FIX] 로그인 이후 화면전환 메서드명 수정

SangJLee1103 committed:
[FIX] 로그인 성공 여부, 사용자 정보 조회 여부에 대한 유즈케이스 프로퍼티 타입 수정

SangJLee1103 committed:
[FIX] 애플 로그인 로직 수정

dlrjswns committed:
[CHORE] SignIn 및 AppCoordinator 코드 수정

- GraceLogAppCoordinator removeChildCoordinator로 수정
- SignInUseCase error Observable 삭제
- SignInReactor 사용안하는 코드 삭제 및 handleSignIn 수정

dlrjswns committed:
[CHORE] SignIn로직 수정
